File: src/main/java/org/academy/internal/client/app/music/backend/decoder/mp3/Mp3AudioStream.java
package org.academy.internal.client.app.music.backend.decoder.mp3;

import javazoom.jl.decoder.*;
import org.academy.AcademyCraft;
import org.academy.internal.client.app.music.backend.decoder.AudioStream;
import org.jspecify.annotations.Nullable;
import org.slf4j.Logger;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ShortBuffer;

public final class Mp3AudioStream implements AudioStream {
    private static final Logger LOGGER = AcademyCraft.getLogger();

    private final ByteBuffer originalData;
    private final int channels;
    private final int sampleRate;
    private final long totalSamples;

    @Nullable
    private Bitstream bitstream;
    @Nullable
    private Decoder decoder;

    @Nullable
    private SampleBuffer currentBuffer;
    private int bufferOffset;

    public Mp3AudioStream(ByteBuffer audioData) throws IOException {
        originalData = audioData;

        var header = readFirstFrame(audioData);
        channels = header.mode() == Header.SINGLE_CHANNEL ? 1 : 2;
        sampleRate = header.frequency();

        var frameCount = header.maxNumberOfFrames(audioData.remaining());
        totalSamples = frameCount * (long) getFrameSampleCount(header);

        resetState();
    }

    private static Header readFirstFrame(ByteBuffer audioData) throws IOException {
        try (var tempStream = new ByteBufferInputStream(audioData.duplicate())) {
            var scanBs = new Bitstream(tempStream);
            var header = scanBs.readFrame();
            scanBs.close();
            if (header == null) throw new IOException("Empty MP3 stream");
            return header;
        } catch (BitstreamException e) {
            throw new IOException("Failed to read MP3 header", e);
        }
    }

    @Override
    public int getChannels() {
        return channels;
    }

    @Override
    public int getSampleRate() {
        return sampleRate;
    }

    @Override
    public long getTotalSamples() {
        return totalSamples;
    }

    @Override
    public int read(ShortBuffer dest) {
        if (bitstream == null || decoder == null) return -1;

        var startPos = dest.position();
        while (
                dest.hasRemaining()
                        && (
                        currentBuffer != null
                                && bufferOffset < currentBuffer.getBufferLength()
                                || decodeNextFrame(bitstream, decoder)
                )
        ) transferTo(dest, currentBuffer);

        var totalWritten = dest.position() - startPos;
        return totalWritten > 0 ? totalWritten / channels : -1;
    }

    private boolean decodeNextFrame(Bitstream bitstream, Decoder decoder) {
        try {
            var header = bitstream.readFrame();
            if (header == null) return false;

            currentBuffer = (SampleBuffer) decoder.decodeFrame(header, bitstream);
            bufferOffset = 0;
            bitstream.closeFrame();
            return true;
        } catch (Exception e) {
            LOGGER.error("MP3 decode error", e);
            return false;
        }
    }

    private void transferTo(ShortBuffer dest, SampleBuffer currentBuffer) {
        var available = currentBuffer.getBufferLength() - bufferOffset;
        var toCopy = Math.min(dest.remaining(), available);
        dest.put(currentBuffer.getBuffer(), bufferOffset, toCopy);
        bufferOffset += toCopy;
    }

    @Override
    public void seek(long targetSample) throws IOException {
        resetState();
        var bs = bitstream;
        var dec = decoder;

        if (bs == null || dec == null) return;

        try {
            advanceToSample(bs, dec, targetSample);
        } catch (BitstreamException | DecoderException e) {
            throw new IOException("Failed to seek in MP3 stream", e);
        }
    }

    private void advanceToSample(Bitstream bs, Decoder dec, long targetSample) throws BitstreamException, DecoderException {
        long currentSample = 0;
        while (currentSample < targetSample) {
            var header = bs.readFrame();
            if (header == null) break;

            var frameLen = getFrameSampleCount(header);
            if (currentSample + frameLen > targetSample) {
                currentBuffer = (SampleBuffer) dec.decodeFrame(header, bs);
                bufferOffset = (int) (targetSample - currentSample);
                bs.closeFrame();
                break;
            }

            currentSample += frameLen;
            bs.closeFrame();
        }
    }

    private void resetState() {
        closeBitstream();
        decoder = new Decoder();
        bitstream = new Bitstream(new ByteBufferInputStream(originalData.duplicate()));
        currentBuffer = null;
        bufferOffset = 0;
    }

    private void closeBitstream() {
        if (bitstream != null) {
            try {
                bitstream.close();
            } catch (BitstreamException e) {
                LOGGER.error("Error closing bitstream", e);
            }
            bitstream = null;
        }
    }

    private static int getFrameSampleCount(Header h) {
        return h.layer() == 1 ? 384 : (h.layer() == 2 ? 1152 : (h.version() == Header.MPEG1 ? 1152 : 576));
    }

    @Override
    public void close() {
        closeBitstream();
        decoder = null;
    }
}

File: src/main/kotlin/org/academy/internal/client/app/music/decoder/mp3/Mp3AudioStream.kt
package org.academy.internal.client.app.music.decoder.mp3

import javazoom.jl.decoder.*
import org.academy.AcademyCraft
import org.academy.internal.client.app.music.decoder.AudioStream
import org.academy.internal.client.app.music.decoder.ByteBufferInputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ShortBuffer

class Mp3AudioStream(audioData: ByteBuffer) : AudioStream {
    private val originalData: ByteBuffer = audioData
    override val channels: Int
    override val sampleRate: Int
    override val totalSamples: Long

    private var bitstream: Bitstream? = null
    private var decoder: Decoder? = null
    private var currentBuffer: SampleBuffer? = null
    private var bufferOffset = 0

    init {
        val header = readFirstFrame(audioData)
        channels = if (header.mode() == Header.SINGLE_CHANNEL) 1 else 2
        sampleRate = header.frequency()
        val frameCount = header.maxNumberOfFrames(audioData.remaining())
        totalSamples = frameCount * getFrameSampleCount(header).toLong()
        resetState()
    }

    private fun readFirstFrame(audioData: ByteBuffer): Header {
        ByteBufferInputStream(audioData.duplicate()).use { tempStream ->
            val scanBs = Bitstream(tempStream)
            val header = scanBs.readFrame()
            scanBs.close()
            if (header == null) throw IOException("Empty MP3 stream")
            return header
        }
    }

    override fun read(pcmBuffer: ShortBuffer): Int {
        val stream = bitstream
        val decoder = decoder
        if (stream == null || decoder == null) return -1
        val startPos = pcmBuffer.position()
        val buffer = currentBuffer
        while (pcmBuffer.hasRemaining() && buffer != null &&
            (bufferOffset < buffer.bufferLength || decodeNextFrame(stream, decoder))
        ) {
            transferTo(pcmBuffer, buffer)
        }
        val totalWritten = pcmBuffer.position() - startPos
        return if (totalWritten > 0) totalWritten / channels else -1
    }

    private fun decodeNextFrame(bitstream: Bitstream, decoder: Decoder): Boolean {
        try {
            val header = bitstream.readFrame() ?: return false
            currentBuffer = decoder.decodeFrame(header, bitstream) as SampleBuffer
            bufferOffset = 0
            bitstream.closeFrame()
            return true
        } catch (e: Exception) {
            logger.error("MP3 decode error", e)
            return false
        }
    }

    private fun transferTo(pcmBuffer: ShortBuffer, currentBuffer: SampleBuffer) {
        val available = currentBuffer.bufferLength - bufferOffset
        val toCopy = minOf(pcmBuffer.remaining(), available)
        pcmBuffer.put(currentBuffer.buffer, bufferOffset, toCopy)
        bufferOffset += toCopy
    }

    override fun seek(sampleFrame: Long) {
        resetState()
        val bs = bitstream
        val dec = decoder
        if (bs == null || dec == null) return
        try {
            advanceToSample(bs, dec, sampleFrame)
        } catch (e: BitstreamException) {
            throw IOException("Failed to seek in MP3 stream", e)
        } catch (e: DecoderException) {
            throw IOException("Failed to seek in MP3 stream", e)
        }
    }

    private fun advanceToSample(bs: Bitstream, dec: Decoder, targetSample: Long) {
        var currentSample = 0L
        while (currentSample < targetSample) {
            val header = bs.readFrame() ?: break
            val frameLen = getFrameSampleCount(header).toLong()
            if (currentSample + frameLen > targetSample) {
                currentBuffer = dec.decodeFrame(header, bs) as SampleBuffer
                bufferOffset = (targetSample - currentSample).toInt()
                bs.closeFrame()
                break
            }
            currentSample += frameLen
            bs.closeFrame()
        }
    }

    private fun resetState() {
        closeBitstream()
        decoder = Decoder()
        bitstream = Bitstream(ByteBufferInputStream(originalData.duplicate()))
        currentBuffer = null
        bufferOffset = 0
    }

    private fun closeBitstream() {
        bitstream?.let {
            try {
                it.close()
            } catch (e: BitstreamException) {
                logger.error("Error closing bitstream", e)
            }
            bitstream = null
        }
    }

    private fun getFrameSampleCount(h: Header): Int =
        when (h.layer()) {
            1 -> 384
            2 -> 1152
            else -> if (h.version() == Header.MPEG1) 1152 else 576
        }

    override fun close() {
        closeBitstream()
        decoder = null
    }

    companion object {
        private val logger = AcademyCraft.getLogger()
    }
}